src/rgw/rgw_sal.cc

```diff
@@ -90,149 +65,46 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
+                                                     bool run_sync_thread,
+                                                     bool run_reshard_thread,
+                                                     bool use_cache,
+                                                     bool use_gc)
 {
-  rgw::sal::Store* store{nullptr};
-
-  if (cfg.store_name.compare("rados") == 0) {
-    store = newStore();
-    RGWRados* rados = static_cast<rgw::sal::RadosStore* >(store)->getRados();
-
-    if ((*rados).set_use_cache(use_cache)
-                .set_use_datacache(false)
-                .set_use_gc(use_gc)
-                .set_run_gc_thread(use_gc_thread)
-                .set_run_lc_thread(use_lc_thread)
-                .set_run_quota_threads(quota_threads)
-                .set_run_sync_thread(run_sync_thread)
-                .set_run_reshard_thread(run_reshard_thread)
-                .init_begin(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    if (store->initialize(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    if (rados->init_complete(dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-  }
-  else if (cfg.store_name.compare("d3n") == 0) {
-    store = new rgw::sal::RadosStore();
-    RGWRados* rados = new D3nRGWDataCache<RGWRados>;
-    dynamic_cast<rgw::sal::RadosStore*>(store)->setRados(rados);
-    rados->set_store(static_cast<rgw::sal::RadosStore* >(store));
-
-    if ((*rados).set_use_cache(use_cache)
-                .set_use_datacache(true)
-                .set_run_gc_thread(use_gc_thread)
-                .set_run_lc_thread(use_lc_thread)
-                .set_run_quota_threads(quota_threads)
-                .set_run_sync_thread(run_sync_thread)
-                .set_run_reshard_thread(run_reshard_thread)
-                .init_begin(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    if (store->initialize(cct, dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-    if (rados->init_complete(dpp) < 0) {
-      delete store;
-      return nullptr;
-    }
-
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_local_datacache_enabled=" <<
-      cct->_conf->rgw_d3n_l1_local_datacache_enabled << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_datacache_persistent_path='" <<
-      cct->_conf->rgw_d3n_l1_datacache_persistent_path << "'" << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_datacache_size=" <<
-      cct->_conf->rgw_d3n_l1_datacache_size << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_evict_cache_on_start=" <<
-      cct->_conf->rgw_d3n_l1_evict_cache_on_start << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_fadvise=" <<
-      cct->_conf->rgw_d3n_l1_fadvise << dendl;
-    lsubdout(cct, rgw, 1) << "rgw_d3n: rgw_d3n_l1_eviction_policy=" <<
-      cct->_conf->rgw_d3n_l1_eviction_policy << dendl;
+  const char *dlname = "/usr/lib64/ceph/librgw_sal_rados.so";
```

Comment:
why is the full path hardcoded?
could it be that the plugins will be in a different directory?
e.g. with vstart, windows machines, etc.

Reply:
because we don't yet have any config for which plug-ins, where I expect the FQpath to be defined.

This is just a placeholder for now

Reply:
and it's a work in progress. :-)

Reply:
I'm told there's some rudimentary config support now. My next push will use that.

Reply:
ceph already has some [plugin support](https://github.com/ceph/ceph/blob/b4d1a19/src/common/PluginRegistry.cc#L135-L143) that we might take advantage of here, whether in full or in part. you can find an example where rgw uses crypto plugins [here in rgw_crypt.cc](https://github.com/ceph/ceph/blob/b4d1a19/src/rgw/rgw_crypt.cc#L291-L298)

`PluginRegistry::load()` refers to a [plugin_dir](https://github.com/ceph/ceph/blob/b4d1a19/src/common/options/global.yaml.in#L720-L724) config variable whose default value comes from our cmake variable `set(CEPH_INSTALL_FULL_PKGLIBDIR ${CMAKE_INSTALL_FULL_LIBDIR}/${PROJECT_NAME})` where `PROJECT_NAME` is `ceph` and `CMAKE_INSTALL_FULL_LIBDIR` comes from https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html

so i think we should look for sal drivers in some subdirectory of `plugin_dir`